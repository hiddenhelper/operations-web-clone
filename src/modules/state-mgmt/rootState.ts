import { History } from 'history';

import { IState as IAuthState } from './auth/state';
import { IState as IGeneralState } from './general/state';
import { IState as IClientState } from './client/state';
import { IState as IProjectState } from './project/state';
import { IState as IUserState } from './user/state';
import { IState as IAccessControlSystemState } from './access-control-system/state';
import { IState as IBadgePrintingSystemState } from './badge-printer-system/state';
import { IState as IWorkerState } from './worker/state';
import { IState as ICertificationState } from './certification/state';
import { IState as IFileState } from './file/state';
import { IState as IBadgeState } from './badge/state';
import { IState as IStatisticsState } from './statistics/state';
import { IState as ITrainingState } from './training/state';
import { IState as IPaymentState } from './payment/state';
import { IState as IInvoiceState } from './invoice/state';
import { ApiService } from '../services/ApiService';
import { AuthService } from '../services/AuthService';
import { PrinterService } from '../services/PrinterService';
import { FileService } from '../services/FileService';
import { Logger } from '../services/Logger';
import { IProcoreState } from './procore/state';
import { IState as IProjectNewState } from './project-new/state';

export interface IAction {
  type: string;
  payload: any;
}

export interface IRootState {
  auth: IAuthState;
  user: IUserState;
  client: IClientState;
  project: IProjectState;
  accessControlSystem: IAccessControlSystemState;
  badgePrinterSystem: IBadgePrintingSystemState;
  worker: IWorkerState;
  certification: ICertificationState;
  training: ITrainingState;
  invoice: IInvoiceState;
  file: IFileState;
  badge: IBadgeState;
  statistics: IStatisticsState;
  payment: IPaymentState;
  general: IGeneralState;
  router?: { location: { pathname: string; search: string; hard: string; key: string }; action: string };
  procore: IProcoreState;
  projectNew: IProjectNewState;
}

export interface IEpicDependencies {
  history: History;
  apiService: ApiService;
  authService: AuthService;
  printerService: PrinterService;
  fileService: FileService;
  logger: typeof Logger;
}
