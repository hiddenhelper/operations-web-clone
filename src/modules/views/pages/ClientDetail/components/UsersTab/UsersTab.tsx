import React, { memo, useEffect, useMemo } from 'react';

import Box from '@material-ui/core/Box';
import Typography from '@material-ui/core/Typography';
import Table from '@material-ui/core/Table';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import TableBody from '@material-ui/core/TableBody';
import Divider from '@material-ui/core/Divider';
import PersonIcon from '@material-ui/icons/Person';
import Avatar from '@material-ui/core/Avatar';
import withStyles from '@material-ui/core/styles/withStyles';

import MenuPopover from '../../../../shared/MenuPopover/MenuPopover';
import Pagination from '../../../../shared/Pagination/Pagination';
import EmptyList from '../../../../shared/EmptyList/EmptyList';

import { UserModel, GeneralModel } from '../../../../../models';
import { formatPhoneNumber, getDefaultValue } from '../../../../../../utils/generalUtils';
import { ProjectsIcon } from '../../../../../../constants';
import { tableGlobalStyles } from '../../../../../../assets/styles';
import { useStyles, tableRowStyles } from '../../styles';

const StyledTableRow = withStyles(tableRowStyles)(TableRow);

export interface IUsersTabProps {
  clientId: string;
  userMap: GeneralModel.IEntityMap<GeneralModel.IEntityMap<UserModel.IUser>>;
  userCount: number;
  queryParams: GeneralModel.IQueryParams;
  loading: GeneralModel.ILoadingStatus;
  onPageChange: (page: number) => void;
  fetchUserList: (id: string, pageNumber: number, pageSize: number) => void;
}

const UsersTab = ({ clientId, userMap, queryParams, userCount, loading, fetchUserList, onPageChange }: IUsersTabProps) => {
  const classes = useStyles();
  const tableGlobalClasses = tableGlobalStyles();

  const userList = useMemo(() => (clientId && Object.keys(userMap).length && userMap[clientId] ? Object.values(userMap[clientId]) : []), [userMap, clientId]);

  const pageCount = useMemo(() => Math.ceil(userCount / queryParams.pageSize), [userCount, queryParams.pageSize]);

  useEffect(() => {
    fetchUserList(clientId, queryParams.pageNumber, queryParams.pageSize);
  }, [clientId, queryParams, fetchUserList]);

  if (loading && loading.isLoading) {
    return <>Loading...</>;
  }
  return (
    <>
      {userList.length === 0 ? (
        <EmptyList icon={<ProjectsIcon />} text="There are no Users assigned" />
      ) : (
        <>
          <div className={tableGlobalClasses.filterActionsContainer}>
            <Box className={`${tableGlobalClasses.filterStatusContainer}`}>
              <Typography className={tableGlobalClasses.filterTextColor}>Show all Projects</Typography>
              <div className={tableGlobalClasses.filterWrapper}>
                <Divider orientation="vertical" flexItem={true} className={tableGlobalClasses.dividerNoSpacing} />
                <span className={tableGlobalClasses.dropdownIcon}>
                  <MenuPopover menuOptionList={[{ title: 'Opt1' }, { title: 'Opt2' }]} placement={'bottom-end'} />
                </span>
              </div>
            </Box>
          </div>
          <Table aria-label="user-list">
            <TableHead>
              <TableRow>
                <TableCell>Name</TableCell>
                <TableCell>Title</TableCell>
                <TableCell>Email</TableCell>
                <TableCell>Mobile Phone</TableCell>
                <TableCell>Office Phone</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              {userList.map(user => (
                <StyledTableRow data-testid="user-client-list-row" key={user.id}>
                  <TableCell className={tableGlobalClasses.cellAvatar}>
                    <Avatar>
                      <PersonIcon />
                    </Avatar>
                    <span>
                      {user.firstName} {user.lastName}
                    </span>
                  </TableCell>
                  <TableCell>{getDefaultValue(user.title)}</TableCell>
                  <TableCell className={classes.boldAccentAnchor}>
                    <a href={`mailto:${user.email}`}>{user.email}</a>
                  </TableCell>
                  <TableCell>{formatPhoneNumber(user.mobilePhoneNumber)}</TableCell>
                  <TableCell>{formatPhoneNumber(user.officePhoneNumber)}</TableCell>
                </StyledTableRow>
              ))}
            </TableBody>
          </Table>
          <Pagination page={queryParams.pageNumber} count={pageCount} onChange={onPageChange} />
        </>
      )}
    </>
  );
};

export default memo(UsersTab);
