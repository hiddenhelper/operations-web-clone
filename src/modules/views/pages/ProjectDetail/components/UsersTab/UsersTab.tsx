import React, { memo, useEffect, useState, useMemo } from 'react';
import { Avatar, Badge, Button, Table, TableHead, TableRow, TableCell, TableBody, withStyles } from '@material-ui/core';
import { Person } from '@material-ui/icons';

import ClientFilter from 'modules/views/shared/ClientFilter';
import EmptyList from 'modules/views/shared/EmptyList';
import Pagination from 'modules/views/shared/Pagination';
import RoleGuard from 'modules/views/shared/RoleGuard';
import TableCellLink from 'modules/views/shared/TableCellLink';

import AssignUser from '../AssignUser';

import { ClientModel, GeneralModel, UserModel } from 'modules/models';
import { AcceptedIcon, ExpiredIcon, NotInvitedIcon, PendingIcon, WorkersIcon } from 'constants/index';
import { getConditionalDefaultValue } from 'utils';

import { avatarGlobalStyles, tableGlobalStyles } from 'assets/styles';
import { useStyles as buttonStyles } from '../../../../shared/FormHandler/ControlledButton/styles';
import { tableRowStyles, useStyles } from '../../styles';

const StyledTableRow = withStyles(tableRowStyles)(TableRow);

export interface IUsersTabProps {
  userMap: GeneralModel.IEntityMap<GeneralModel.IEntityMap<UserModel.IUserProject>>;
  clientMap?: GeneralModel.IEntityMap<GeneralModel.IEntityMap<ClientModel.IClientProject>>;
  userCount: number;
  queryParams: GeneralModel.IQueryParams;
  isModalOpen: boolean;
  projectId: string;
  ctaDisabled: boolean;
  isFcAdmin: boolean;
  userLoading: GeneralModel.ILoadingStatus;
  assignUserLoading: GeneralModel.ILoadingStatus;
  openModal: () => void;
  closeModal: () => void;
  onPageChange: (page: number) => void;
  clearUserMap: () => void;
  fetchUserList: (id: string, query: GeneralModel.IQueryParams) => void;
  fetchProjectClientList?: (id: string, query: GeneralModel.IQueryParams) => void;
  setQueryParams: (params) => void;
}

const UsersTab = ({
  userMap,
  userCount,
  queryParams,
  isModalOpen,
  projectId,
  clientMap,
  ctaDisabled,
  isFcAdmin,
  assignUserLoading,
  userLoading,
  fetchUserList,
  openModal,
  closeModal,
  onPageChange,
  clearUserMap,
  setQueryParams,
  fetchProjectClientList,
}: IUsersTabProps) => {
  const classes = useStyles();
  const avatarGlobalClasses = avatarGlobalStyles();
  const tableGlobalClasses = tableGlobalStyles();
  const buttonClasses = buttonStyles();

  const [isEditUser, setIsEditUser] = useState<boolean>(false);

  const countUsers = useMemo(() => Math.ceil(userCount / queryParams.limit), [userCount, queryParams.limit]);

  const userList = useMemo(() => (projectId && Object.keys(userMap).length && userMap[projectId] ? Object.values(userMap[projectId]) : []), [
    userMap,
    projectId,
  ]);

  const userStatusIcon = useMemo(
    () => ({
      [UserModel.InvitationStatus.ACCEPTED]: <AcceptedIcon />,
      [UserModel.InvitationStatus.PENDING]: <PendingIcon />,
      [UserModel.InvitationStatus.NOT_INVITED]: <NotInvitedIcon />,
      [UserModel.InvitationStatus.EXPIRED]: <ExpiredIcon />,
    }),
    []
  );

  useEffect(() => {
    /* istanbul ignore else */
    if (assignUserLoading && !assignUserLoading.isLoading) {
      closeModal();
      fetchUserList(projectId, queryParams);
    }
  }, [assignUserLoading, projectId, queryParams, closeModal, fetchUserList]);

  useEffect(() => {
    fetchUserList(projectId, queryParams);
  }, [projectId, queryParams, fetchUserList]);

  useEffect(() => {
    return function unMount() {
      clearUserMap();
    };
  }, [clearUserMap]);

  if (userLoading && userLoading.isLoading) {
    return <>Loading...</>;
  }

  const openEditTab = () => {
    if (!isFcAdmin) return;

    setIsEditUser(true);
    openModal();
  };

  const onCloseModal = () => {
    setIsEditUser(false);
    closeModal();
  };

  return (
    <>
      <div className={`${tableGlobalClasses.filterActionsContainer} ${tableGlobalClasses.filterActionsContainerPadding}`}>
        <ClientFilter
          queryParams={queryParams}
          clientMap={clientMap[projectId]}
          projectId={projectId}
          isFcAdmin={isFcAdmin}
          setQueryParams={setQueryParams}
          fetchClientList={fetchProjectClientList}
        />
        <Button
          className={`${buttonClasses.createButton} ${buttonClasses.primaryButtonLarge}`}
          color="primary"
          variant="contained"
          fullWidth={true}
          size="large"
          type="submit"
          data-testid="user-modal-open-btn"
          onClick={openModal}
          disabled={ctaDisabled}
        >
          Assign User
        </Button>
      </div>
      {userList.length === 0 ? (
        <EmptyList data-testid="empty-user-list" icon={<WorkersIcon />} text="There are no Users assigned" />
      ) : (
        <>
          <div className={classes.userTableWrapper}>
            <Table aria-label="user-list">
              <TableHead>
                <TableRow>
                  <TableCell>Name</TableCell>
                  <RoleGuard roleList={[UserModel.Role.FCA_ADMIN]}>
                    <>
                      <TableCell>Type</TableCell>
                      <TableCell>Title</TableCell>
                      <TableCell>Client</TableCell>
                      <TableCell>Project Role</TableCell>
                    </>
                  </RoleGuard>
                  <RoleGuard roleList={[UserModel.Role.CLIENT_ADMIN, UserModel.Role.REGULAR_USER]}>
                    <>
                      <TableCell>Title</TableCell>
                      <TableCell>Project Role</TableCell>
                      <TableCell>Company</TableCell>
                    </>
                  </RoleGuard>
                </TableRow>
              </TableHead>
              <TableBody>
                {userList.map(user => (
                  <StyledTableRow data-testid="client-list-row" key={user.id}>
                    <TableCell>
                      <div className={classes.userCenter}>
                        <Badge
                          className={`${classes.avatarPosition} ${avatarGlobalClasses.avatarWithSmallBadge}`}
                          overlap="circle"
                          anchorOrigin={{
                            vertical: 'bottom',
                            horizontal: 'right',
                          }}
                          badgeContent={userStatusIcon[user.invitationStatus]}
                        >
                          <Avatar className={`${avatarGlobalClasses.avatarWrapper} ${avatarGlobalClasses.avatarMedium}`}>
                            <Person titleAccess={UserModel.userInviteMap[user.invitationType]} />
                          </Avatar>
                        </Badge>
                        <span onClick={openEditTab} className={classes.userTitle}>
                          {user.firstName} {user.lastName}
                        </span>
                      </div>
                    </TableCell>
                    <RoleGuard roleList={[UserModel.Role.FCA_ADMIN]}>
                      <TableCell>{UserModel.userInviteMap[user.invitationType]}</TableCell>
                    </RoleGuard>
                    <TableCell>{user.title}</TableCell>
                    <RoleGuard roleList={[UserModel.Role.FCA_ADMIN]}>
                      <TableCell>
                        {user.company?.id ? (
                          <TableCellLink href={`/clients/detail/${user.company.id}`} text={user.company.name} title="View Client details" />
                        ) : (
                          '-'
                        )}
                      </TableCell>
                    </RoleGuard>
                    <TableCell>{getConditionalDefaultValue(user.roleName, `${user.roleName}`, 'Not defined')}</TableCell>
                    <RoleGuard roleList={[UserModel.Role.CLIENT_ADMIN, UserModel.Role.REGULAR_USER]}>
                      <TableCell>{user.company?.name}</TableCell>
                    </RoleGuard>
                  </StyledTableRow>
                ))}
              </TableBody>
            </Table>
          </div>
          <Pagination page={queryParams.page} count={countUsers} onChange={onPageChange} />
        </>
      )}

      {isModalOpen && <AssignUser id={projectId} closeModal={onCloseModal} isFcAdmin={isFcAdmin} isEditUser={isEditUser} />}
    </>
  );
};

export default memo(UsersTab);
