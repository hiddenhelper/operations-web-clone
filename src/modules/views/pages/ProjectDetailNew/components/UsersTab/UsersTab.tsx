import React, { memo, useEffect, useState, useMemo } from 'react';
import { Avatar, Badge, Button, Table, TableHead, TableRow, TableCell, TableBody, withStyles } from '@material-ui/core';
import { Person } from '@material-ui/icons';

import ClientFilter from 'modules/views/shared/ClientFilter';
import EmptyList from 'modules/views/shared/EmptyList';
import Pagination from 'modules/views/shared/Pagination';
import PermissionGuard from 'modules/views/shared/PermissionGuard';
import TableCellLink from 'modules/views/shared/TableCellLink';

import AssignUser from '../AssignUser';

import { ClientModel, GeneralModel, UserModel } from 'modules/models';
import { AcceptedIcon, ExpiredIcon, NotInvitedIcon, PendingIcon, WorkersIcon } from 'constants/index';
import { getConditionalDefaultValue } from 'utils';

import { avatarGlobalStyles, tableGlobalStyles } from 'assets/styles';
import { useStyles as buttonStyles } from '../../../../shared/FormHandler/ControlledButton/styles';
import { tableRowStyles, useStyles } from '../../styles';

const StyledTableRow = withStyles(tableRowStyles)(TableRow);

export interface IUsersTabProps {
  userMap: GeneralModel.IEntityMap<GeneralModel.IEntityMap<UserModel.IUserProject>>;
  clientMap?: GeneralModel.IEntityMap<GeneralModel.IEntityMap<ClientModel.IClientProject>>;
  userCount: number;
  queryParams: GeneralModel.IQueryParams;
  isModalOpen: boolean;
  projectId: string;
  ctaDisabled: boolean;
  isFcaUser: boolean;
  userLoading: GeneralModel.ILoadingStatus;
  assignUserLoading: GeneralModel.ILoadingStatus;
  openModal: () => void;
  closeModal: () => void;
  onPageChange: (page: number) => void;
  clearUserMap: () => void;
  fetchUserList: (id: string, query: GeneralModel.IQueryParams) => void;
  fetchProjectClientList?: (id: string, query: GeneralModel.IQueryParams) => void;
  setQueryParams: (params) => void;
  fetchUserProfile: (companyId: string, companyUserId: string) => void;
}

const UsersTab = ({
  userMap,
  userCount,
  queryParams,
  isModalOpen,
  projectId,
  clientMap,
  ctaDisabled,
  isFcaUser,
  assignUserLoading,
  userLoading,
  fetchUserList,
  openModal,
  closeModal,
  onPageChange,
  clearUserMap,
  setQueryParams,
  fetchProjectClientList,
  fetchUserProfile,
}: IUsersTabProps) => {
  const classes = useStyles();
  const avatarGlobalClasses = avatarGlobalStyles();
  const tableGlobalClasses = tableGlobalStyles();
  const buttonClasses = buttonStyles();

  const [isEditUser, setIsEditUser] = useState<boolean>(false);

  const [userCompanyId, setUserCompanyId] = useState<string>('');

  const countUsers = useMemo(() => Math.ceil(userCount / queryParams.limit), [userCount, queryParams.limit]);

  const userList = useMemo(() => (projectId && Object.keys(userMap).length && userMap[projectId] ? Object.values(userMap[projectId]) : []), [
    userMap,
    projectId,
  ]);

  const userStatusIcon = useMemo(
    () => ({
      [UserModel.InvitationStatus.ACCEPTED]: <AcceptedIcon />,
      [UserModel.InvitationStatus.PENDING]: <PendingIcon />,
      [UserModel.InvitationStatus.NOT_INVITED]: <NotInvitedIcon />,
      [UserModel.InvitationStatus.EXPIRED]: <ExpiredIcon />,
    }),
    []
  );

  useEffect(() => {
    /* istanbul ignore else */
    if (assignUserLoading && !assignUserLoading.isLoading) {
      closeModal();
      fetchUserList(projectId, queryParams);
    }
  }, [assignUserLoading, projectId, queryParams, closeModal, fetchUserList]);

  useEffect(() => {
    fetchUserList(projectId, queryParams);
  }, [projectId, queryParams, fetchUserList]);

  useEffect(() => {
    return function unMount() {
      clearUserMap();
    };
  }, [clearUserMap]);

  if (userLoading && userLoading.isLoading) {
    return <>Loading...</>;
  }

  const editCompanyUser = (companyId: string, companyUserId: string) => {
    if (!companyId || !companyUserId) return;

    fetchUserProfile(companyId, companyUserId);
    setUserCompanyId(companyId);
    setIsEditUser(true);
    openModal();
  };

  const onCloseModal = () => {
    fetchUserList(projectId, queryParams);
    setIsEditUser(false);
    closeModal();
  };

  return (
    <>
      <div className={`${tableGlobalClasses.filterActionsContainer} ${tableGlobalClasses.filterActionsContainerPadding}`}>
        <ClientFilter
          queryParams={queryParams}
          clientMap={clientMap[projectId]}
          projectId={projectId}
          isFcaUser={isFcaUser}
          setQueryParams={setQueryParams}
          fetchClientList={fetchProjectClientList}
        />
        <PermissionGuard permissionsExpression={`${UserModel.UsersPermission.MANAGE} AND ${UserModel.ProjectsPermission.VIEWACCESS}`}>
          <Button
            className={`${buttonClasses.createButton} ${buttonClasses.primaryButtonLarge}`}
            color="primary"
            variant="contained"
            fullWidth={true}
            size="large"
            type="submit"
            data-testid="user-modal-open-btn"
            onClick={openModal}
            disabled={ctaDisabled}
          >
            Assign User
          </Button>
        </PermissionGuard>
      </div>
      {userList.length === 0 ? (
        <EmptyList data-testid="empty-user-list" icon={<WorkersIcon />} text="There are no Users assigned" />
      ) : (
        <>
          <div className={classes.userTableWrapper}>
            <Table aria-label="user-list">
              <TableHead>
                <TableRow>
                  <TableCell>Name</TableCell>
                  <TableCell>Title</TableCell>
                  {isFcaUser && <TableCell>Client</TableCell>}
                  <TableCell>Project Role</TableCell>
                  {!isFcaUser && <TableCell>Company</TableCell>}
                </TableRow>
              </TableHead>
              <TableBody>
                {userList.map(user => (
                  <StyledTableRow data-testid="client-list-row" key={user.id}>
                    <TableCell>
                      <div className={classes.userCenter}>
                        <Badge
                          className={`${classes.avatarPosition} ${avatarGlobalClasses.avatarWithSmallBadge}`}
                          overlap="circle"
                          anchorOrigin={{
                            vertical: 'bottom',
                            horizontal: 'right',
                          }}
                          badgeContent={userStatusIcon[user.invitationStatus]}
                        >
                          <Avatar className={`${avatarGlobalClasses.avatarWrapper} ${avatarGlobalClasses.avatarMedium}`}>
                            <Person titleAccess={UserModel.userInviteMap[user.invitationType]} />
                          </Avatar>
                        </Badge>
                        <PermissionGuard
                          permissionsExpression={`${UserModel.UsersPermission.VIEWACCESS} AND ${UserModel.UsersPermission.MANAGE}`}
                          fallback={
                            <span className={classes.userTitle}>
                              {user.firstName} {user.lastName}
                            </span>
                          }
                        >
                          <span
                            onClick={() => (user.invitationStatus === UserModel.InvitationStatus.PENDING ? editCompanyUser(user.company.id, user.id) : {})}
                            className={classes.userTitle}
                          >
                            {user.firstName} {user.lastName}
                          </span>
                        </PermissionGuard>
                      </div>
                    </TableCell>
                    <TableCell>{user.title || '-'}</TableCell>
                    {isFcaUser && (
                      <TableCell>
                        {user.company?.id ? (
                          <PermissionGuard permissionsExpression={UserModel.ClientsPermission.VIEWACCESS} fallback={<>{user.company.name}</>}>
                            <TableCellLink href={`/clients/detail/${user.company.id}`} text={user.company.name} title="View Client details" />
                          </PermissionGuard>
                        ) : (
                          '-'
                        )}
                      </TableCell>
                    )}
                    <TableCell>{getConditionalDefaultValue(user.roleName, `${user.roleName}`, 'Not defined')}</TableCell>
                    {!isFcaUser && <TableCell>{user.company?.name}</TableCell>}
                  </StyledTableRow>
                ))}
              </TableBody>
            </Table>
          </div>
          <Pagination page={queryParams.page} count={countUsers} onChange={onPageChange} />
        </>
      )}

      {isModalOpen && <AssignUser id={projectId} closeModal={onCloseModal} isEditUser={isEditUser} companyId={userCompanyId} />}
    </>
  );
};

export default memo(UsersTab);
