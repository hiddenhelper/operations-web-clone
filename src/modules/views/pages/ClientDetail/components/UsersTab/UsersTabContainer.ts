import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import UsersTab from './UsersTab';

import { GENERAL } from '../../../../../../constants';
import { IRootState } from '../../../../../state-mgmt/rootState';
import { userState } from '../../../../../state-mgmt/user';

export const mapStateToProps = (state: IRootState) => ({
  userMap: state.user.userClientMap,
  userCount: state.user.count,
  loading: state.general.loadingMap[GENERAL.LOADING_KEY.FETCH_CLIENT_USER_LIST],
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchUserList: (id: string, pageNumber: number, pageSize: number) => dispatch(userState.actions.fetchClientUserListStart(id, pageNumber, pageSize)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UsersTab);
