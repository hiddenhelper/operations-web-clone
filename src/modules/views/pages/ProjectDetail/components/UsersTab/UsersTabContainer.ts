import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import UsersTab from './UsersTab';

import { GeneralModel } from '../../../../../models';
import { GENERAL } from '../../../../../../constants';
import { IRootState } from '../../../../../state-mgmt/rootState';
import { userState } from '../../../../../state-mgmt/user';
import { clientState } from '../../../../../state-mgmt/client';

export const mapStateToProps = (state: IRootState) => ({
  userMap: state.user.userProjectMap,
  clientMap: state.client.clientProjectMap,
  userCount: state.user.count,
  userLoading: state.general.loadingMap[GENERAL.LOADING_KEY.FETCH_PROJECT_USER_LIST],
  assignUserLoading: state.general.loadingMap[GENERAL.LOADING_KEY.ASSIGN_USER_PROJECT],
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchUserList: (id: string, query: GeneralModel.IQueryParams) => dispatch(userState.actions.fetchProjectUserListStart(id, query)),
  fetchProjectClientList: (id: string, query: GeneralModel.IQueryParams) => dispatch(clientState.actions.fetchProjectClientListStart(id, query)),
  clearUserMap: () => dispatch(userState.actions.clearUserMap()),
  fetchUserProfile: (companyId: string, companyUserId: string) => dispatch(userState.actions.fetchUserProfileStart(companyId, companyUserId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UsersTab);
