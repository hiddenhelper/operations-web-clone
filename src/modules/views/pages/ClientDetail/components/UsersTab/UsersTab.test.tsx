import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import { createMockStore } from 'redux-test-utils';
import { render } from '@testing-library/react';

import UsersTab, { IUsersTabProps } from './UsersTab';
import { getClient_1, getUser_1 } from '../../../../../../test/entities';
import { getInitialState } from '../../../../../../test/rootState';

describe.skip('UsersTab', () => {
  let props: IUsersTabProps;

  beforeEach(() => {
    props = {
      userCount: 1,
      userMap: {
        [getClient_1().id]: {
          [getUser_1().id]: { ...getUser_1() },
        },
      },
      loading: {
        isLoading: false,
        hasError: false,
        error: null,
      },
      queryParams: { page: 1, limit: 30 },
      clientId: getClient_1().id,
      onPageChange: jest.fn(),
      fetchUserList: jest.fn(),
    };
  });

  it.skip('should render empty', () => {
    props.userMap = {};
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <UsersTab {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.container).toMatchSnapshot();
    expect(wrapper.getByText('There are no Users assigned'));
  });

  it('should render loading', () => {
    props.loading = { isLoading: true, hasError: false, error: null };
    const wrapper = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <UsersTab {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it.skip('should render', () => {
    const { container } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <UsersTab {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should fetchUserList', () => {
    props.userMap = {};
    render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <UsersTab {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(props.fetchUserList).toHaveBeenCalled();
  });
});
