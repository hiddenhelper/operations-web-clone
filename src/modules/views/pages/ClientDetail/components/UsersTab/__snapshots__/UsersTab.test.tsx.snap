// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTab should render 1`] = `
<div>
  <div
    class="makeStyles-filterActionsContainer-139"
  >
    <div
      class="MuiBox-root MuiBox-root-167 makeStyles-filterStatusContainer-142"
    >
      <p
        class="MuiTypography-root makeStyles-filterTextColor-150 MuiTypography-body1"
      >
        Show all Projects
      </p>
      <div
        class="makeStyles-filterWrapper-147"
      >
        <hr
          class="MuiDivider-root makeStyles-dividerNoSpacing-153 MuiDivider-flexItem MuiDivider-vertical"
        />
        <span
          class="makeStyles-dropdownIcon-151"
        >
          <div
            class="makeStyles-iconInactive-168"
            data-testid="popover-wrapper"
          >
            <span
              class=""
              data-testid="controlled-button"
              tabindex="-1"
            >
              <button
                aria-haspopup="true"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="popover-button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg>
                    expand.svg
                  </svg>
                </span>
              </button>
            </span>
            
          </div>
        </span>
      </div>
    </div>
  </div>
  <table
    aria-label="user-list"
    class="MuiTable-root"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Name
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Title
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Email
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Mobile Phone
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Office Phone
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root WithStyles(ForwardRef(TableRow))-root-172"
        data-testid="user-client-list-row"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-cellAvatar-154"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
          <span>
            Pedro
             
            Martin
          </span>
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          -
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-boldAccentAnchor-123"
        >
          <a
            href="mailto:user@test.com"
          >
            user@test.com
          </a>
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          -
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          -
        </td>
      </tr>
    </tbody>
  </table>
  <nav
    aria-label="pagination navigation"
    class="MuiPagination-root  makeStyles-pagination-173"
    data-testid="pagination"
  >
    <ul
      class="MuiPagination-ul"
    >
      <li>
        <button
          aria-label="Go to previous page"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiPaginationItem-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </button>
      </li>
      <li>
        <button
          aria-label="Go to next page"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiPaginationItem-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`UsersTab should render empty 1`] = `
<div>
  <div
    class=" makeStyles-container-50"
  >
    <div>
      <div
        class="MuiAvatar-root MuiAvatar-circle makeStyles-missingAvatarWrapper-61 makeStyles-missingAvatarBig-62 MuiAvatar-colorDefault"
      >
        <svg>
          projects.svg
        </svg>
      </div>
    </div>
    <p
      class="MuiTypography-root makeStyles-missingListText-51 MuiTypography-body1"
    >
      There are no Users assigned
    </p>
  </div>
</div>
`;

exports[`UsersTab should render loading 1`] = `
<div>
  Loading...
</div>
`;
